Userland/Libraries/LibCore/Stream.h

```diff
@@ -672,67 +696,63 @@ class BufferedHelper {
-                    longest_match = result;
-                    match_size = candidate.length();
+            for (auto& [span, offset] : spans_and_offsets) {
+                auto result = AK::memmem_optional(span.data(), span.size(), candidate.bytes().data(), candidate.bytes().size());
+                if (result.has_value()) {
+                    *result += offset;
+                    auto previous_match = longest_match.value_or(*result);
+                    if ((previous_match < *result) || (previous_match == *result && match_size < candidate.length())) {
+                        longest_match = result;
+                        match_size = candidate.length();
+                    }
                 }
             }
         }
         if (longest_match.has_value()) {
-            auto size_written_to_user_buffer = *longest_match;
-            auto buffer_to_take = m_buffer.span().slice(0, size_written_to_user_buffer);
-            auto buffer_to_shift = m_buffer.span().slice(size_written_to_user_buffer + match_size);
-
-            buffer_to_take.copy_to(buffer);
-            m_buffer.overwrite(0, buffer_to_shift.data(), buffer_to_shift.size());
-
-            m_buffered_size -= size_written_to_user_buffer + match_size;
-
-            return buffer.slice(0, size_written_to_user_buffer);
+            auto const buffer_written_for_end_user = m_buffer.read(buffer.slice(0, *longest_match));
+            m_buffer.discard(match_size);
+            return buffer_written_for_end_user;
         }
 
         // If we still haven't found anything, then it's most likely the case
         // that the delimiter ends beyond the length of the caller-passed
         // buffer. Let's just fill the caller's buffer up.
-        auto buffer_to_take = m_buffer.span().slice(0, readable_size);
-        auto buffer_to_shift = m_buffer.span().slice(readable_size);
-
-        buffer_to_take.copy_to(buffer);
-        m_buffer.overwrite(0, buffer_to_shift.data(), buffer_to_shift.size());
-
-        m_buffered_size -= readable_size;
-
-        return buffer.slice(0, readable_size);
+        return m_buffer.read(buffer);
     }
 
     // Returns whether a line can be read, populating the buffer in the process.
     ErrorOr<bool> can_read_line()
     {
-        if (stream().is_eof() && m_buffered_size > 0)
+        if (stream().is_eof() && m_buffer.size() > 0)
             return true;
 
-        if (m_buffer.span().slice(0, m_buffered_size).contains_slow('\n'))
-            return true;
+        for (auto const span : m_buffer.each_separate_span()) {
+            if (span.contains_slow('\n'))
+                return true;
+        }
 
         if (!stream().is_readable())
             return false;
 
-        while (m_buffered_size < m_buffer.size()) {
-            auto populated_slice = TRY(populate_read_buffer());
+        while (m_buffer.size() < m_buffer.capacity()) {
+            auto const populated_slices = TRY(populate_read_buffer());
```

Comment:
Do we even have to return the actual slices here? We could just make it tell us how much data it read, and then start searching the buffer at an offset using the public interface (see previous comment). That would drop a lot of logic both from here and from `populate_read_buffer`.